modules/core/src/weaver/Expectations.scala

```diff
@@ -211,12 +211,45 @@ object Expectations {
         pos: SourceLocation): Expectations =
       Expectations.Additive.unwrap(la.foldMap(a => Expectations.Additive(f(a))))
 
+    /**
+     * Checks that a given expression matches a certain pattern; fails
+     * otherwise.
+     *
+     * @example
+     *   {{{
+     *     matches(Option(4) { case Some(x) =>
+     *       expect.eql(4, x)
+     *     }
+     *   }}}
+     */
+    def matches[A](x: A)(
+        f: PartialFunction[A, Expectations]
+    )(
+        implicit pos: SourceLocation,
+        A: Show[A] = Show.fromToString[A]): Expectations =
+      if (f.isDefinedAt(x))
+        f(x)
+      else
+        failure("Pattern did not match, got: " + A.show(x))
+
     /**
      * Alias for `forEach`
      */
     def inEach[L[_], A](la: L[A])(f: A => Expectations)(
         implicit L: Foldable[L]): Expectations = forEach(la)(f)
 
+    def existsRight[F[_], A, E](fa: F[A])(
+        f: A => Expectations
+    )(
+        implicit pos: SourceLocation,
+        F: ApplicativeError[F, E],
+        G: Foldable[F],
+        E: Show[E] = Show.fromToString[E]): Expectations =
+      fa
+        .map(f)
+        .handleError(e => failure(e.show))
+        .foldLeft(failure("impossible?"))((_, x) => x)
```

Comment:
May be `.attempt.map(_.fold(...))`?

Reply:
Wasn't totally sure what the goal here would be. Doesn't something need to interpret `F`? In the PR's case, that's what `F : Foldable` is for

`attempt` would still be wrapped in `F`